### DISCUSSION TITLE
how to understand the world and local in the Jacobian

### DISCUSSION THREAD

**Discussion by LHT-Liu** [stack-of-tasks/pinocchio#2191] [Q&A] [answered]
![image](https://github.com/stack-of-tasks/pinocchio/assets/64388319/a3b9ea0a-4961-4191-a778-7d803fd95b32)

Hello,i can get jacobian by computeJointJacobians or computeJointJacobians, and i see [the message ](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#abadfcd21d13fa17865a189e4b47acca9)that in the world frame, data.J means what (last joint relative to 0 joint（universe joint）)?,but i can also get jacobian by pin.getJointJacobian(model, data, Joint7, pin.LOCAL),where i don't understand the input pin.local or pin.world，who relative to who? Finally pin.computeJointJacobian(model, data, q_random, Joint7) doesn't have the ref, which means the input ref joint relative to 0 joint （universe Joint） ?

thanks

**Comment by florent-lamiraux** [comment, marked as answer]
Hello @LIUHAITAO-CH 
We modified the documentation you display in your message very recently. Please refer to the new version in [this discussion](https://github.com/stack-of-tasks/pinocchio/discussions/2183) and let us know if it is clearer for you.

**Reply by LHT-Liu** [reply]
![image](https://github.com/stack-of-tasks/pinocchio/assets/64388319/0229e6ca-610d-4648-95a9-21688c6de511)
As the picture shows, As the picture shows, there is a problem with the representation or rendering of mathematical symbols

**Reply by fabinsch** [reply]
thanks for pointing this out. You are right that currently there are some `\f` missing in front of the `$` for the math rendering. We will fix it soon. If you fix it locally and generate the docs you will see the following

![image](https://github.com/stack-of-tasks/pinocchio/assets/55981657/87510f85-95f2-4d28-ba6d-f52b767c74af)

**Reply by LHT-Liu** [reply]
hello，the following is what i understand

```
import pinocchio as pin
from pinocchio.visualize import MeshcatVisualizer
import time
import numpy as np

urdf_path = "../urdf/Franka_urdf/fr3_without_hand.urdf"
model: pin.Model = pin.buildModelFromUrdf(urdf_path)

data = model.createData()

q_random = pin.randomConfiguration(model)

Joint7 = model.getJointId("fr3_joint7")

end_frame = model.getFrameId("fr3_joint8")

pin.computeJointJacobians(model, data, q_random)

Joint7_jacobian = pin.getJointJacobian(model, data, Joint7, pin.WORLD)

end_jacobian = pin.getFrameJacobian(model, data, end_frame, pin.WORLD)

Joint7_jacobian_ = pin.computeJointJacobian(model, data, q_random, Joint7)
```
this is my code.

I think 

- `Joint7_jacobian`,(joint7 w.r.t. joint0)

- `Joint7_jacobian_` ,(joint7 w.r.t. joint7,itself, )

- and There is no `Reference Frame` in the function input, which is relative to its local coordinate system，such as `pin.computeFrameJacobian`

is  the above right?

**Reply by fabinsch** [reply]
yes, this is right. As stated in the docs for [computeJointJacobian](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#af1948cb50a9211fd338feb2a3d42d661), it returns 
>the Jacobian of the specific joint frame expressed in the local frame of the joint (matrix 6 x model.nv).

It says in the remark that 
>The result of this function is equivalent to call first computeJointJacobians(model,data,q) and then call getJointJacobian(model,data,jointId,LOCAL,J)

Here you have a snapshot of the improved documentation for `ReferenceFrame`. This was merged into devel in #2190 but still not into master, therefore you are not seeing it online. But you can generate it locally.
![image](https://github.com/stack-of-tasks/pinocchio/assets/55981657/3481eb1b-77b9-4e84-923e-344e65e90587)

**Reply by LHT-Liu** [reply]
ok,thaks for your reply. 

Can I get the Jacobian matrix of joint 7 relative to joint 2? Is there a similar function?

**Comment by traversaro** [comment]
In case this may be useful to someone, back in time to simplify using pinocchio in the lab I am part of (https://github.com/ami-iit), I wrote a document that goes a bit more in the mathematical details of `WORLD`/`LOCAL`/`LOCAL_WORLD_ALIGNED`, mainly from the point of view of the notation (https://research.tue.nl/en/publications/multibody-dynamics-notation-version-2) and existing libraries (iDynTree, https://github.com/robotology/idyntree) that the users in the lab are familiar with. For most of it I checked my assumptions with tests, so hopefully the theory is correct, even if obviously they may be errors.

The document is available in https://github.com/ami-iit/idynfor/blob/master/doc/theory_background.md, probably it is of limited use if you are not familiar with the notation it is used in there (that is different in some aspects from the Featherstone one), but anyhow I guess it may be useful for someone.
